File: ExtremeCore/Controls/PathTextBox.vb
Imports System.Windows.Forms

Public Class PathTextBox

    Public Enum PathTypes
        Folder
        FileSave
        FileOpen
    End Enum
    Public Property PathType As PathTypes

    ''' <summary>
    ''' Will be used for .Description in folders and as .Title for Files.
    ''' </summary>
    ''' <returns></returns>
    Public Property Description As String

    ''' <summary>
    ''' Only useable if PathType is set to FileSave or FileOpen
    ''' </summary>
    ''' <returns></returns>
    Public Property Filter As String

    Private Sub BrowseBtn_Click(sender As Object, e As EventArgs) Handles BrowseBtn.Click
        If PathType = PathTypes.Folder Then
            Dim dlg As New FolderBrowserDialog()
            dlg.Description = Description
            If dlg.ShowDialog() = DialogResult.OK Then
                PathText.Text = dlg.SelectedPath
            End If

        ElseIf PathType = PathTypes.FileOpen Then
            Dim dlg As New OpenFileDialog()
            dlg.Title = Description
            dlg.Filter = Filter
            If dlg.ShowDialog() = DialogResult.OK Then
                PathText.Text = dlg.FileName
            End If

        ElseIf PathType = PathTypes.FileSave Then
            Dim dlg As New SaveFileDialog()
            dlg.Title = Description
            dlg.Filter = Filter
            If dlg.ShowDialog() = DialogResult.OK Then
                PathText.Text = dlg.FileName
            End If

        End If
    End Sub

    Private Sub PathText_MouseDoubleClick(sender As Object, e As Windows.Forms.MouseEventArgs) Handles PathText.MouseDoubleClick
        BrowseBtn.PerformClick()
    End Sub
End Class

File: ExtremeCore/Controls And Tools/PathTextBox.cs
using System;
using System.Windows.Forms;

namespace ExtremeCore.Controls_And_Tools
{
    public partial class PathTextBox : UserControl
    {
        public PathTextBox()
        {
            InitializeComponent();
        }

        public enum PathTypes
        {
            Folder,
            FileSave,
            FileOpen
        }
        public PathTypes PathType {get; set;}
	
        /// <summary>
        /// Will be used for .Description in folders and as .Title for Files.
        /// </summary>
        /// <returns></returns>
        public string Description {get; set;}
	
        /// <summary>
        /// Only useable if PathType is set to FileSave or FileOpen
        /// </summary>
        /// <returns></returns>
        public string Filter {get; set;}
	
        private void BrowseBtn_Click(object sender, EventArgs e)
        {
            if (PathType == PathTypes.Folder)
            {
                FolderBrowserDialog dlg = new FolderBrowserDialog
                {
                    Description = Description
                };
                if (dlg.ShowDialog() == DialogResult.OK)
                {
                    PathText.Text = dlg.SelectedPath;
                }
			
            }
            else if (PathType == PathTypes.FileOpen)
            {
                OpenFileDialog dlg = new OpenFileDialog
                {
                    Title = Description,
                    Filter = Filter
                };
                if (dlg.ShowDialog() == DialogResult.OK)
                {
                    PathText.Text = dlg.FileName;
                }
			
            }
            else if (PathType == PathTypes.FileSave)
            {
                SaveFileDialog dlg = new SaveFileDialog
                {
                    Title = Description,
                    Filter = Filter
                };
                if (dlg.ShowDialog() == DialogResult.OK)
                {
                    PathText.Text = dlg.FileName;
                }
			
            }
        }
	
        private void PathText_MouseDoubleClick(object sender, MouseEventArgs e)
        {
            BrowseBtn.PerformClick();
        }
    }
}
